Add a getting started guide on workload qualification (#4110)
NVIDIA/spark-rapids: docs/get-started/getting-started-workload-qualification.md

---
layout: page
title: Workload Qualification
nav_order: 7
parent: Getting-Started
---

> # Getting Started on Spark workload qualification

Author: Yes I plan to add a link from getting started main page to this doc so that people know where to start.

> # Getting Started on Spark workload qualification

Author: @tgravescs  I just added a link in `getting-stated.md`:

```
## Spark workload qualification

If you plan to convert existing Spark workload from CPU to GPU, please refer to this
[Spark workload qualification](./getting-started-workload-qualification.md) to check if your Spark Applications are good
fit for the RAPIDS Accelerator for Apache Spark.
```

The RAPIDS Accelerator for Apache Spark does not support all the features which are supported in Apache Spark.
If you plan to convert existing Spark workload from CPU to GPU, it is highly recommended that you do gap analysis 
to figure out if there are any unsupported features such as functions, expressions, data types, data formats.
After that we will know which workload is more suitable to GPU, which is called "workload qualification" here.

If there are certain operators which can not run on GPU due to the current limitations, they will fallback to CPU mode.

> As a result, it may incur some performance overhead because of host memory <=> GPU memory transfer. 

Reviewer: Probably should mention that it is not critical to ensure all operations are supported by the GPU and significant performance benefits can be gained before all operations are fully supported by the GPU.  It all depends on how critical the portion that is executing on the CPU is to the overall performance of the query.  For example, if most of the query is on the GPU and a hash aggregation towards the end of the query ends up being on the CPU but only a few thousand rows end up going through the CPU processing, it's not going to be super critical to ensure that last aggregation gets on the GPU because it will contribute so little to the overall query performance.  Obviously running the query with the RAIPDS Accelerator is best, but use the relative row counts and/or data sizes moving through various stages of the query to help estimate the relative importance of having an operation be GPU accelerated or not.

Author: I added below words from @jlowe 's suggestion:
```
However it is not critical to ensure all operations are supported by the GPU and significant performance benefits can 
be gained before all operations are fully supported by the GPU. It all depends on how critical the portion that is 
executing on the CPU is to the overall performance of the query.
```
Is that enough?

Reviewer: Adding that text should be fine to cover it.

This article is to help you get familiar with the best practice and different tools we provide on how to do gap analysis 
and workload qualification.

> 

Author: I see, instead of putting a picture, we can have some decision tree right? We can decide after that discussion.

## 1. Qualification and Profiling tool

### Requirement

- Access to Spark event logs from Spark 2.x or 3.x
- Spark 3.0.1+ jars
- `rapids-4-spark-tools` jar

### How to use

> If you have old Spark event logs from Spark 2.x or 3.x, you can use [Qualification tool](../spark-qualification-tool.md) 

Reviewer 2: perhaps just old to be something more along the lines of Spark event logs from prior runs of the application. the event logs used might only be minutes "old".

Add the to the part: "you can use the Qualification tool]

Author: Changed to:
```
If you have Spark event logs from prior runs of the applications on Spark 2.x or 3.x, you can use the
[Qualification tool]
```
How do you think?

Reviewer 2: that works

and [Profiling tool](../spark-profiling-tool.md) to analyze them.

> Qualification tool can output the score, rank and some of the potential not-supported features for each Spark application.

Reviewer: ```suggestion
The qualification tool outputs the score, rank and some of the potentially not-supported features for each Spark application.
```

Its output can help you focus on the top N Spark applications which are SQL heavy applications. 

> Profiling tool can output SQL plan metrics and also print out actual query plans to provide more insights.

Reviewer: ```suggestion
The profiling tool outputs SQL plan metrics and also prints out actual query plans to provide more insights.
```

> For example, below Profiling tool output for a specific Spark application shows that it has a query with a large 

Reviewer: ```suggestion
The following example profiling tool output for a specific Spark application shows that it has a query with a large 
```

> `HashAggregate` and `SortMergeJoin`. Those are indicators for a good candidate for RAPIDS accelerator.

Reviewer: ```suggestion
`HashAggregate` and `SortMergeJoin`. Those are indicators for a good candidate for the RAPIDS Accelerator.
```

```
+--------+-----+------+----------------------------------------------------+-------------+------------------------------------+-------------+----------+
|appIndex|sqlID|nodeID|nodeName                                            |accumulatorId|name                                |max_value    |metricType|
+--------+-----+------+----------------------------------------------------+-------------+------------------------------------+-------------+----------+
|1       |88   |8     |SortMergeJoin                                       |11111        |number of output rows               |500000000    |sum       |
|1       |88   |9     |HashAggregate                                       |22222        |number of output rows               |600000000    |sum       |
```

> Since the 2 tools are basically analyzing Spark event logs, it can not provide very accurate gap analysis information.
> But it is very convenient because you do not need a CPU or GPU Spark cluster to run them.

Reviewer: ```suggestion
Since the two tools are only analyzing Spark event logs, it can not provide very accurate gap analysis information.
However it is very convenient because you do not need a CPU or GPU Spark cluster to run them.
```

## 2. Function `explainPotentialGPUPlan` 

### Requirement

- A Spark 3.x CPU cluster
- A pair of `rapids-4-spark` and `cudf` jars

### How to use

> In Spark RAPIDS accelerator 21.12, a new function named `explainPotentialGPUPlan` is added which can help us understand 

Reviewer: ```suggestion
Starting with version 21.12 of the RAPIDS Accelerator, a new function named `explainPotentialGPUPlan` is added which can help us understand 
```

the potential GPU plan and if there are any not-supported features on a CPU cluster.
Basically it can print the logs which are the same as the driver logs with `spark.rapids.sql.explain=all`.

1. In `spark-shell`, add the `rapids-4-spark` and `cudf` jars into --jars option or put them in Spark related classpath.

   For example:

   ```bash
   spark-shell --jars /PathTo/cudf-<version>.jar,/PathTo/rapids-4-spark_<version>.jar
   ```

2. Test if the class can be successfully loaded or not.

   ```scala
   import com.nvidia.spark.rapids.ExplainPlan.explainPotentialGPUPlan
   ```

> 3. Enable Spark RAPIDS related parameters to allow supporting existing features.

Reviewer: ```suggestion
3. Enable RAPIDS Accelerator related parameters to allow supporting existing features.
```

> 3. Enable Spark RAPIDS related parameters to allow supporting existing features.

Author: Below is what i added as the disclaimer:
```
   Please refer to [config doc](../configs.md) for details of RAPIDS Accelerator parameters.
   Note: Please understand the meaning and risk of above parameters before enabling it. 
```
How is that ?

> 3. Enable Spark RAPIDS related parameters to allow supporting existing features.

Reviewer 2: yeah that sounds better.  I do still wonder about changing the title of the step, maybe something like:

Enable optional RAPIDS Accelerator parameters
Enable optional RAPIDS Accelerator parameters based to your setup

I do wonder about saying something like below as the first sentence in the section
Enabling optional parameter may allow more operations to run on the GPU but please understand the ....

> 3. Enable Spark RAPIDS related parameters to allow supporting existing features.

Author: I changed it to below in the beginning:
```
   Enabling optional parameters may allow more operations to run on the GPU but please understand the meaning and risk 
   of above parameters before enabling it. Please refer to [configs doc](../configs.md) for details of RAPIDS 
   Accelerator parameters.
```

>    For example, if your jobs have `double`/`float`/`decimal` operators together with some scala UDFs, you can set 
>    below parameters:

Reviewer: ```suggestion
   For example, if your jobs have `double`/`float`/`decimal` operators together with some Scala UDFs, you can set 
   the following parameters:
```

   
   ```scala
   spark.conf.set("spark.rapids.sql.incompatibleOps.enabled", true)
   spark.conf.set("spark.rapids.sql.variableFloatAgg.enabled", true)
   spark.conf.set("spark.rapids.sql.decimalType.enabled", true)
   spark.conf.set("spark.rapids.sql.castFloatToDecimal.enabled",true)
   spark.conf.set("spark.rapids.sql.castDecimalToFloat.enabled",true)
   spark.conf.set("spark.rapids.sql.udfCompiler.enabled",true)
   ```
   

>    Please refer to [config doc](../configs.md) for details of those Spark RAPIDS related parameters.

Reviewer: ```suggestion
   Please refer to [config doc](../configs.md) for details of RAPIDS Accelerator parameters.
```

> 4. Run the function `explainPotentialGPUPlan` on top of Dataframe.

Reviewer: ```suggestion
4. Run the function `explainPotentialGPUPlan` on the query DataFrame.
```

   For example:

   ```scala
   scala> val df_multi=spark.sql("SELECT value82*value63 FROM df2 union SELECT value82+value63 FROM df2")
   df_multi: org.apache.spark.sql.DataFrame = [(CAST(value82 AS DECIMAL(9,3)) * CAST(value63 AS DECIMAL(9,3))): decimal(15,5)]

   scala> val output=com.nvidia.spark.rapids.ExplainPlan.explainPotentialGPUPlan(df_multi)
   scala> println(output)
   ```

   Below are sample driver log messages starting with `!` which indicate the unsupported features in this version:
   
   ```
   !Exec <ProjectExec> cannot run on GPU because not all expressions can be replaced
     !Expression <Multiply> (promote_precision(cast(value82#30 as decimal(9,3))) * promote_precision(cast(value63#31 as decimal(9,3)))) cannot run on GPU because The actual output precision of the multiply ilarge to fit on the GPU DecimalType(19,6)
   ```

This log can show you which operators(on what data type) can not run on GPU and what is the reason.

> If it shows a specific Spark RAPIDS parameter which can be turned on to enable that feature, you can enable that parameter 

Reviewer: ```suggestion
If it shows a specific RAPIDS Accelerator parameter which can be turned on to enable that feature, you can enable that parameter 
```

> and try the tool again.

Reviewer: Probably want to mention here that one shouldn't blindly enable parameters that are disabled by default without understanding the risks and applicability of that parameters ramifications to their specific application.

Author: I added some words to ask users to understand the risks of the parameters. How is now?

Reviewer: This is better.  Would be nice to link to the configs.md document here so users can know where to lookup any parameters they may find in the output.

Author: Added the configs doc link here.

`HashAggregate` and `SortMergeJoin`. Those are indicators for a good candidate for the RAPIDS Accelerator.

```
+--------+-----+------+----------------------------------------------------+-------------+------------------------------------+-------------+----------+
|appIndex|sqlID|nodeID|nodeName                                            |accumulatorId|name                                |max_value    |metricType|
+--------+-----+------+----------------------------------------------------+-------------+------------------------------------+-------------+----------+
|1       |88   |8     |SortMergeJoin                                       |11111        |number of output rows               |500000000    |sum       |
|1       |88   |9     |HashAggregate                                       |22222        |number of output rows               |600000000    |sum       |
```

Since the two tools are only analyzing Spark event logs, it can not provide very accurate gap analysis information.
However it is very convenient because you do not need a CPU or GPU Spark cluster to run them.

> 

Reviewer 2: so the other thing that I think is missing is really how the user interprets this output.  Again this kind of goes along with Chris' doc perhaps but they run this tool then what?  I'm fine with that being separate and perhaps going into each of the docs, but we may also want to say something like,... run qualification tool get score, if things look ok, then run profiling tool to get query plans, or skip to explainPotentialGPUPlan if possible, etc..  If we aren't doing that here can you please file a separate issue to follow up once we have discussed with Chris and team.

Author: For profiling tool, i gave one example on how to interprets the output such as looking at SQL plan metrics.
For qualification tool, how about we just mention the `unsupported data types`, `udf` etc which could be related to gap analysis?

Reviewer 2: yeah that's fine for now.

Author: I just added:
```
The qualification tool outputs the score, rank and some of the potentially not-supported features for each Spark application.
For example, the CSV output can output `Potential Problems`, `Unsupported Read File Formats and Types`, and
`Unsupported Write Data Format` which are the indication of some not-supported features.
Its output can help you focus on the top N Spark applications which are SQL heavy applications.
```

## 2. Function `explainPotentialGPUPlan` 

### Requirement

- A Spark 3.x CPU cluster
- A pair of `rapids-4-spark` and `cudf` jars

> 

Reviewer 2: maybe we should put modify your code to call up here

Author: Added:
```
- Need to modify the existing Spark application code

```

   scala> val output=com.nvidia.spark.rapids.ExplainPlan.explainPotentialGPUPlan(df_multi)
   scala> println(output)
   ```

   Below are sample driver log messages starting with `!` which indicate the unsupported features in this version:
   
   ```
   !Exec <ProjectExec> cannot run on GPU because not all expressions can be replaced
     !Expression <Multiply> (promote_precision(cast(value82#30 as decimal(9,3))) * promote_precision(cast(value63#31 as decimal(9,3)))) cannot run on GPU because The actual output precision of the multiply ilarge to fit on the GPU DecimalType(19,6)
   ```

This log can show you which operators(on what data type) can not run on GPU and what is the reason.
If it shows a specific RAPIDS Accelerator parameter which can be turned on to enable that feature, you should firstly 
understand the risk and applicability of that parameter based on [configs doc](../configs.md) and then enable that 
parameter and try the tool again.

Since its output is directly based on specific version of `rapids-4-spark` jar, the gap analysis is pretty accurate.

> But you need a Spark 3+ CPU cluster and is ok to modify the code to add this function.

Reviewer 2: maybe we can remove this line if the requirement section covers

Author: Removed.

`HashAggregate` and `SortMergeJoin`. Those are indicators for a good candidate for the RAPIDS Accelerator.

```
+--------+-----+------+----------------------------------------------------+-------------+------------------------------------+-------------+----------+
|appIndex|sqlID|nodeID|nodeName                                            |accumulatorId|name                                |max_value    |metricType|
+--------+-----+------+----------------------------------------------------+-------------+------------------------------------+-------------+----------+
|1       |88   |8     |SortMergeJoin                                       |11111        |number of output rows               |500000000    |sum       |
|1       |88   |9     |HashAggregate                                       |22222        |number of output rows               |600000000    |sum       |
```

Since the two tools are only analyzing Spark event logs, it can not provide very accurate gap analysis information.
However it is very convenient because you do not need a CPU or GPU Spark cluster to run them.

## 2. Function `explainPotentialGPUPlan` 

> ### Requirement
> 
> - A Spark 3.x CPU cluster
> - A pair of `rapids-4-spark` and `cudf` jars
> - Need to modify the existing Spark application code

Reviewer 3: Formatting and rewording
```suggestion
### Requirements

- A Spark 3.x CPU cluster
- A pair of `rapids-4-spark` and `cudf` jars
- Ability to modify the existing Spark application code
```

### How to use

Starting with version 21.12 of the RAPIDS Accelerator, a new function named `explainPotentialGPUPlan` is added which can help us understand 
the potential GPU plan and if there are any not-supported features on a CPU cluster.
Basically it can return output which is the same as the driver logs with `spark.rapids.sql.explain=all`.

1. In `spark-shell`, add the `rapids-4-spark` and `cudf` jars into --jars option or put them in Spark related classpath.

   For example:

   ```bash
   spark-shell --jars /PathTo/cudf-<version>.jar,/PathTo/rapids-4-spark_<version>.jar
   ```

2. Test if the class can be successfully loaded or not.

   ```scala
   import com.nvidia.spark.rapids.ExplainPlan.explainPotentialGPUPlan
   ```

3. Enable optional RAPIDS Accelerator related parameters based on your setup.

>    Enabling optional parameters may allow more operations to run on the GPU but please understand the meaning and risk 
>    of above parameters before enabling it. Please refer to [configs doc](../configs.md) for details of RAPIDS 
>    Accelerator parameters.
>    
>    For example, if your jobs have `double`/`float`/`decimal` operators together with some Scala UDFs, you can set 
>    the following parameters:

Reviewer 3: Formatting and minor rewording
```suggestion
   Enabling optional parameters may allow more operations to run on the GPU but please understand
   the meaning and risk of above parameters before enabling it. Please refer to [configs
   doc](../configs.md) for details of RAPIDS Accelerator parameters.
   
   For example, if your jobs have `double`, `float` and `decimal` operators together with some Scala
   UDFs, you can set the following parameters:
```

   
   ```scala
   spark.conf.set("spark.rapids.sql.incompatibleOps.enabled", true)
   spark.conf.set("spark.rapids.sql.variableFloatAgg.enabled", true)
   spark.conf.set("spark.rapids.sql.decimalType.enabled", true)
   spark.conf.set("spark.rapids.sql.castFloatToDecimal.enabled",true)
   spark.conf.set("spark.rapids.sql.castDecimalToFloat.enabled",true)
   spark.conf.set("spark.rapids.sql.udfCompiler.enabled",true)
   ```

4. Run the function `explainPotentialGPUPlan` on the query DataFrame.

   For example:

   ```scala
   scala> val df_multi=spark.sql("SELECT value82*value63 FROM df2 union SELECT value82+value63 FROM df2")
   df_multi: org.apache.spark.sql.DataFrame = [(CAST(value82 AS DECIMAL(9,3)) * CAST(value63 AS DECIMAL(9,3))): decimal(15,5)]

   scala> val output=com.nvidia.spark.rapids.ExplainPlan.explainPotentialGPUPlan(df_multi)
   scala> println(output)
   ```

>    Below are sample driver log messages starting with `!` which indicate the unsupported features in this version:

Reviewer 3: Formatting
```suggestion
   Below are sample driver log messages starting with `!` which indicate the unsupported features in
   this version:
```

   
   ```
   !Exec <ProjectExec> cannot run on GPU because not all expressions can be replaced
     !Expression <Multiply> (promote_precision(cast(value82#30 as decimal(9,3))) * promote_precision(cast(value63#31 as decimal(9,3)))) cannot run on GPU because The actual output precision of the multiply ilarge to fit on the GPU DecimalType(19,6)
   ```

> This log can show you which operators(on what data type) can not run on GPU and what is the reason.
> If it shows a specific RAPIDS Accelerator parameter which can be turned on to enable that feature, you should firstly 
> understand the risk and applicability of that parameter based on [configs doc](../configs.md) and then enable that 
> parameter and try the tool again.
> 
> Since its output is directly based on specific version of `rapids-4-spark` jar, the gap analysis is pretty accurate.

Reviewer 3: Formatting and rewording
```suggestion
This log can show you which operators (on what data type) can not run on GPU and the reason.
If it shows a specific RAPIDS Accelerator parameter which can be turned on to enable that feature,
you should firstly understand the risk and applicability of that parameter based on [configs
doc](../configs.md) and then enable that parameter and try the tool again.

Since its output is directly based on specific version of `rapids-4-spark` jar, the gap analysis is
pretty accurate.
```

## 3. Run Spark applications with Spark RAPIDS Accelerator on a GPU Spark Cluster

> ### Requirement
> 
> - A Spark 3.x GPU cluster
> - A pair of `rapids-4-spark` and `cudf` jars
> 
> ### How to use
> 
> Follow the getting-started guides to start a Spark 3+ GPU cluster and run the existing Spark workloads on the GPU 
> cluster with parameter `spark.rapids.sql.explain=all`.
> The Spark driver log should be collected to check the not-supported messages.
> This is the most accurate way to do gap analysis.

Reviewer 3: Formatting and rewording
```suggestion
### Requirements

- A Spark 3.x GPU cluster
- The `rapids-4-spark` and `cudf` [jars](../download.md)

### How to use

Follow the getting-started guides to start a Spark 3+ GPU cluster and run the existing Spark
workloads on the GPU cluster with parameter `spark.rapids.sql.explain=all`.  The Spark driver log
should be collected to check the not-supported messages.  This is the most accurate way to do gap
analysis.
```

---
layout: page
title: Workload Qualification
nav_order: 7
parent: Getting-Started
---
# Getting Started on Spark workload qualification

The RAPIDS Accelerator for Apache Spark runs as many operations as possible on the GPU.  If there
are operators which do not yet run on GPU, they will seamlessly fallback to the CPU.  There may be
some performance overhead because of host memory to GPU memory transfer.  When converting an
existing Spark workload from CPU to GPU, it is recommended to do an analysis to understand if there
are any features (functions, expressions, data types, data formats) that do not yet run on the GPU.
Understanding this will help prioritize workloads that are best suited to the GPU.

Significant performance benefits can be gained even if all operations are not yet fully supported by
the GPU. It all depends on how critical the portion that is executing on the CPU is to the overall
performance of the query.

> This article describe the tools we provide and how to do gap analysis and workload qualification.

Reviewer 3: ```suggestion
This article describes the tools we provide and how to do gap analysis and workload qualification.
```

   the meaning and risk of above parameters before enabling it. Please refer to [configs
   doc](../configs.md) for details of RAPIDS Accelerator parameters.
   
   For example, if your jobs have `double`, `float` and `decimal` operators together with some Scala
   UDFs, you can set the following parameters:
   
   ```scala
   spark.conf.set("spark.rapids.sql.incompatibleOps.enabled", true)
   spark.conf.set("spark.rapids.sql.variableFloatAgg.enabled", true)
   spark.conf.set("spark.rapids.sql.decimalType.enabled", true)
   spark.conf.set("spark.rapids.sql.castFloatToDecimal.enabled",true)
   spark.conf.set("spark.rapids.sql.castDecimalToFloat.enabled",true)
   spark.conf.set("spark.rapids.sql.udfCompiler.enabled",true)
   ```

4. Run the function `explainPotentialGPUPlan` on the query DataFrame.

   For example:

   ```scala
   scala> val df_multi=spark.sql("SELECT value82*value63 FROM df2 union SELECT value82+value63 FROM df2")
   df_multi: org.apache.spark.sql.DataFrame = [(CAST(value82 AS DECIMAL(9,3)) * CAST(value63 AS DECIMAL(9,3))): decimal(15,5)]

   scala> val output=com.nvidia.spark.rapids.ExplainPlan.explainPotentialGPUPlan(df_multi)
   scala> println(output)

>    ```

Reviewer 3: Will this example be out of date in 21.12 with decimal 128 support?

Author: yes could be. This is an example based on some un-supported messages.

Author: Or should I mention that:
This is an example missing feature based on 21.10 version but it could be implemented in the future release.

Author: @sameerz I added a JDBC Scan example instead.

   Below are sample driver log messages starting with `!` which indicate the unsupported features in
   this version:
   
   ```
   !Exec <ProjectExec> cannot run on GPU because not all expressions can be replaced
     !Expression <Multiply> (promote_precision(cast(value82#30 as decimal(9,3))) * promote_precision(cast(value63#31 as decimal(9,3)))) cannot run on GPU because The actual output precision of the multiply ilarge to fit on the GPU DecimalType(19,6)
   ```

This log can show you which operators (on what data type) can not run on GPU and the reason.
If it shows a specific RAPIDS Accelerator parameter which can be turned on to enable that feature,

> you should firstly understand the risk and applicability of that parameter based on [configs

Reviewer 3: ```suggestion
you should first understand the risk and applicability of that parameter based on [configs
```

doc](../configs.md) and then enable that parameter and try the tool again.

Since its output is directly based on specific version of `rapids-4-spark` jar, the gap analysis is
pretty accurate.

## 3. Run Spark applications with Spark RAPIDS Accelerator on a GPU Spark Cluster

### Requirements

- A Spark 3.x GPU cluster
- The `rapids-4-spark` and `cudf` [jars](../download.md)

### How to use

Follow the getting-started guides to start a Spark 3+ GPU cluster and run the existing Spark
workloads on the GPU cluster with parameter `spark.rapids.sql.explain=all`.  The Spark driver log
should be collected to check the not-supported messages.  This is the most accurate way to do gap

> analysis.

Reviewer 3: Do we need an example in this section?

Author: I just added an example to show how to understand the not-supported messages. How is now?
